hail/src/main/scala/is/hail/io/RowStore.scala

```diff
@@ -175,6 +231,87 @@ final case class PackCodecSpec(child: BufferSpec) extends CodecSpec {
+          ifDefined.emit(fieldPtr.store(NCode(s"reinterpret_cast<char *>(*reinterpret_cast<long *>($fieldPtr))")))
+          encodeArray(buf, ifDefined, t2, fieldPtr)
+        case _: PBoolean => buf.writeByte(ifDefined, NCode(s"$fieldPtr ? 0 : 1"))
+        case _: PInt32 => buf.writeInt(ifDefined, NCode(s"*reinterpret_cast<int *>($fieldPtr)"))
+        case _: PInt64 => buf.writeLong(ifDefined, NCode(s"*reinterpret_cast<long *>($fieldPtr)"))
+        case _: PFloat32 => buf.writeFloat(ifDefined, NCode(s"*reinterpret_cast<float *>($fieldPtr)"))
+        case _: PFloat64 => buf.writeDouble(ifDefined, NCode(s"*reinterpret_cast<double *>($fieldPtr)"))
+        case _: PBinary =>
+          ifDefined.emit(fieldPtr.store(NCode(s"reinterpret_cast<char *>(*reinterpret_cast<long *>($fieldPtr))")))
+          encodeBinary(buf, ifDefined, fieldPtr)
+      }
+      i += 1
+    }
+  }
+
+  def encodeArray(buf: NativeOutputBuffer, fb: NativeScope, t: PArray, ptr: NativeCodeVariable): Unit = {
+    val length = fb.newRef("length", "unsigned int")
+    fb.emit(length.store(NCode(s"*reinterpret_cast<int *>($ptr)")))
+    buf.writeInt(fb, length.load())
+    if (!t.elementType.required) {
+      val nMissingBytes = fb.newRef("nMissingBytes", "unsigned int")
+      fb.emit(nMissingBytes.store(NCode(s"($length + 7) >> 3")))
+      buf.writeBytes(fb, NCode(s"$ptr + 4"), nMissingBytes.load())
+    }
+    val eltPtr = fb.newRef("eltPtr", "char *")
+    fb.emit(eltPtr.store(NCode(s"$ptr + (${t.elementsOffset(length.load())})")))
+    val i = fb.newRef("i", "int", NCode("0"))
+    val elements = fb._while(NCode(s"$i < $length"))
+    val ifDefined: NativeScope = if (!t.elementType.required) {
+      val newScope = elements._if(t.isElementDefined(ptr.load(), i.load()))
+      newScope
+    } else elements
+
+    t.elementType match {
+      case t2: PBaseStruct => encodeBaseStruct(buf, ifDefined, t2, eltPtr)
+      case t2: PArray =>
+        val eltPtr2 = fb.newRef("eltPtr2", "char *")
+        ifDefined.emit(eltPtr2.store(NCode(s"reinterpret_cast<char *>(*reinterpret_cast<long *>($eltPtr))")))
+        encodeArray(buf, ifDefined, t2, eltPtr2)
+      case _: PBoolean => buf.writeByte(ifDefined, NCode(s"$eltPtr ? 0 : 1"))
+      case _: PInt32 => buf.writeInt(ifDefined, NCode(s"*reinterpret_cast<int *>($eltPtr)"))
+      case _: PInt64 => buf.writeLong(ifDefined, NCode(s"*reinterpret_cast<long *>($eltPtr)"))
+      case _: PFloat32 => buf.writeFloat(ifDefined, NCode(s"*reinterpret_cast<float *>($eltPtr)"))
+      case _: PFloat64 => buf.writeDouble(ifDefined, NCode(s"*reinterpret_cast<double *>($eltPtr)"))
+      case _: PBinary =>
+        val eltPtr2 = fb.newRef("eltPtr2", "char *")
+        ifDefined.emit(eltPtr2.store(NCode(s"reinterpret_cast<char *>(*reinterpret_cast<long *>($eltPtr))")))
+        encodeBinary(buf, ifDefined, eltPtr2)
+    }
+
+    elements.emit(eltPtr.store(NCode(s"$eltPtr + ${t.elementByteSize}")))
+    elements.emit(NCode(s"$i++"))
+
+  }
```

Comment:
So I find this a bit hard to read.  What about:

```
  def encodeArray(buf: NativeOutputBuffer, fb: NativeScope, t: PArray, ptr: NativeCodeVariable): Unit = {
    val length = fb.define("unsigned", "length", "*reinterpret_cast<int *>($ptr)")
    fb += s"$buf.writeInt($length)"

    if (!t.elementType.required)
      fb += s"$buf.writeBytes($ptr + 4, ($length + 7) >> 3);"
    val eltPtr = fb.define("char *", "eltPtr", "$ptr + (${ t.elementsOffset($length) })")
    val i = fb.define("int", "i", "0")

    var emitElem = t.elements match { ... }
    if (!t.elementType.required)
      // or emitElem = If(t.isElementDefined(ptr, i), emitElem)

    fb += """
while ($i < $length) {
  $emitElem
  $eltPtr = $eltPtr + ${ t.elementByteSize }
  ++i;
}
"""
   }
```

I like being able to drop into strings for larger blocks of stuff that are just straight C++.

Another alternative which involves a bit more explicit C++ and less Scala machinery (just for generating symbol names) is:

```
  def encodeArray(buf: NativeOutputBuffer, fb: NativeScope, t: PArray, ptr: NativeCodeVariable): Unit = {
    val length = genSym("length")
    fb += s"""
unsigned length = load_int($ptr);
$buf.writeInt($length);
"""
  ...
}
```

I also think we should have C++ functions like `load_int(char *ptr)` for `*reinterpret_cast<int *>(ptr)`.